### PR TITLE
[IMP] sale,account: Remove the proforma stage form account and add ne…

### PR DESCRIPTION
**Pull request by mag-odoo**
Description of the feature this PR addresses:
Task:https://www.odoo.com/web#id=30137&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad:https://pad.odoo.com/p/r.4de7f8c81dc5a08018d3b7aacc5818e1

Current behavior before PR:
Proforma stage is in account app.

Desired behavior after PR is merged:
Adding new button for sending proforma mail and adding new report for proforma invoice.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_invoice.py [hunk 1]
```diff
@@ -550,16 +548,10 @@ def action_invoice_draft(self):
...
-            raise UserError(_("Invoice must be a draft in order to set it to Pro-forma."))
-        return self.write({'state': 'proforma2'})
-
-    @api.multi
     def action_invoice_open(self):
         # lots of duplicate calls to action_invoice_open, so we remove those already open
         to_open_invoices = self.filtered(lambda inv: inv.state != 'open')
-        if to_open_invoices.filtered(lambda inv: inv.state not in ['proforma2', 'draft']):
+        if to_open_invoices.filtered(lambda inv: inv.state != 'draft'):
             raise UserError(_("Invoice must be in draft or Pro-forma state in order to validate it."))
```
Comment (jir-odoo) [inline addons/account/models/account_invoice.py]: Change error message where it uses pro-forma

File: addons/sale/models/sale_config_settings.py [hunk 2]
```diff
@@ -48,6 +48,8 @@ class SaleConfiguration(models.TransientModel):
         "Show total",
         implied_group='sale.group_show_price_total',
         group='base.group_portal,base.group_user,base.group_public')
+    group_proforma_sales = fields.Boolean(string="Pro-forma", implied_group='sale.group_proforma_sales',
+        help="Allows you to put sales orders in pro-forma state.")
```
Comment (mdi-odoo) [inline addons/sale/models/sale_config_settings.py]: Change **help** message to **Allows you to send pro-forma invoice.**

File: addons/sale/models/sale.py [hunk 3]
```diff
@@ -34,6 +34,39 @@ def _amount_all(self):
                 'amount_total': amount_untaxed + amount_tax,
             })
 
+    @api.multi
+    def action_invoice_proforma(self):
+        self.ensure_one()
+        ir_model_data = self.env['ir.model.data']
+        try:
+            template_id = ir_model_data.get_object_reference('sale', 'email_template_edi_proforma_sale')[1]
```
Comment (mdi-odoo) [inline addons/sale/models/sale.py]: May be you can use **self.env.ref()** .

File: addons/sale/models/sale.py [hunk 4]
```diff
@@ -34,6 +34,39 @@ def _amount_all(self):
...
+        except ValueError:
+            template_id = False
+        try:
+            compose_form_id = ir_model_data.get_object_reference('mail', 'email_compose_message_wizard_form')[1]
```
Comment (mdi-odoo) [inline addons/sale/models/sale.py]: Same as above.

File: addons/sale/views/sale_views.xml [hunk 5]
```diff
@@ -155,9 +155,11 @@
                         type="action" context="{'default_advance_payment_method': 'percentage'}"
                         attrs="{'invisible': ['|',('invoice_status', '!=', 'no'), ('state', '!=', 'sale')]}"/>
                     <button name="action_quotation_send" string="Send by Email" type="object" states="draft" class="btn-primary"/>
+                    <button name="action_invoice_proforma" type="object" string="SEND PRO-FORMA" groups="sale.group_proforma_sales" class="btn-primary" attrs="{'invisible': ['|', ('state', 'in', ['sent', 'sale']), ('invoice_count','&gt;=',1)]}"/>
```
Comment (mdi-odoo) [inline addons/sale/views/sale_views.xml]: May be **SEND PRO-FORMA** button is only visible in **draft** state.

File: addons/sale/views/sale_views.xml [hunk 6]
```diff
@@ -155,9 +155,11 @@
...
                     <button name="print_quotation" string="Print" type="object" states="draft" class="btn-primary"/>
                     <button name="action_confirm" states="sent" string="Confirm Sale" class="btn-primary o_sale_confirm" type="object" />
                     <button name="action_confirm" states="draft" string="Confirm Sale" class="o_sale_confirm" type="object" />
+                    <button name="action_invoice_proforma" type="object" string="SEND PRO-FORMA" groups="sale.group_proforma_sales" attrs="{'invisible': ['|', ('state', '=', 'draft'), ('invoice_count','&gt;=',1)]}"/>
```
Comment (mdi-odoo) [inline addons/sale/views/sale_views.xml]: Why you add same button two times ?

File: addons/sale/views/sale_config_settings_views.xml [hunk 7]
```diff
@@ -246,6 +246,17 @@
...
                         </div>
                     </div>
+                    <div class="col-xs-12 col-md-6 o_setting_box">
+                        <div class="o_setting_left_pane">
+                            <field name="group_proforma_sales"/>
+                        </div>
+                        <div class="o_setting_right_pane">
+                            <label for="group_proforma_sales"/>
+                            <div class="text-muted">
+                                Get a pro-forma status for sales
```
Comment (mdi-odoo) [inline addons/sale/views/sale_config_settings_views.xml]: Rename to **Send a pro-forma invoice to customer**

File: addons/sale/views/sale_config_settings_views.xml [hunk 8]
```diff
@@ -246,6 +246,17 @@
...
                         </div>
                     </div>
+                    <div class="col-xs-12 col-md-6 o_setting_box">
+                        <div class="o_setting_left_pane">
+                            <field name="group_proforma_sales"/>
+                        </div>
+                        <div class="o_setting_right_pane">
+                            <label for="group_proforma_sales"/>
+                            <div class="text-muted">
+                                Send a Pro-Forma to customer
```
Comment (bst-odoo) [inline addons/sale/views/sale_config_settings_views.xml]: Allows you to send Pro-Forma to your customers

File: addons/sale/data/mail_template_data.xml [hunk 9]
```diff
@@ -14,28 +14,38 @@
...
 % set doc_name = 'quotation' if object.state in ('draft', 'sent') else 'order confirmation'
 % set is_online = access_action and access_action['type'] == 'ir.actions.act_url'
 % set access_name = is_online and object.template_id and 'Accept and pay %s online' % doc_name or 'View %s' % doc_name
 % set access_url = is_online and access_action['url'] or ''
 
-% if object.partner_id.parent_id:
-    (<i>${object.partner_id.parent_id.name}</i>)
+<p>Dear
+% if object.partner_id.is_company and object.child_ids:
+    ${object.partner_id.child_ids[0].name}
```
Comment (tivisse) [inline addons/sale/data/mail_template_data.xml]: Why displaying the first child ? Is there a functional reason behind this ?

Comment (mag-odoo) [inline addons/sale/data/mail_template_data.xml]: Due to conflicts resolution , there is no functional reason behind this.

**Comment by tivisse** [comment]
Landed in master at https://github.com/odoo/odoo/commit/c3cf05af3ef9a5055981594163fcfbd71c9be811

**Pull request merged**